nucleus/common/internal-api/src/main/java/org/glassfish/internal/data/EngineInfo.java

```diff
@@ -100,29 +101,35 @@ public void clean(ExtendedDeploymentContext context) throws Exception {
 
     /*
      * Sets the registry this container belongs to
+     *
      * @param the registry owning me
      */
     public void setRegistry(ContainerRegistry registry) {
         this.registry = registry;
     }
 
     // Todo : take care of Deployer when unloading...
-    public void stop(Logger logger)
-    {
-        if (getDeployer()!=null) {
-            ServiceHandle<?> i = registry.habitat.getServiceHandle(getDeployer().getClass());
-            if (i!=null) {
-                i.destroy();
+    public void stop(Logger logger) {
+        if (getDeployer() != null) {
+            ServiceHandle<?> serviceHandle = registry.serviceLocator.getServiceHandle(getDeployer().getClass());
+            if (serviceHandle != null) {
+                serviceHandle.close();
             }
         }
 
         if (container != null && container.isActive()) {
-            container.destroy();
+            container.close();
         }
 
         registry.removeContainer(this);
-        if (logger.isLoggable(Level.FINE)) {
+
+        if (logger.isLoggable(FINE)) {
             logger.fine("Container " + getContainer().getName() + " stopped");
         }
     }
+
+    @Override
+    public String toString() {
+        return (deployer != null? deployer.toString() : "")  + super.toString();
```

Comment:
just replace it with `return deployer + ", " + super.toString();`
Btw super.toString is Object.toString(), is it what is desired?